starknet_py/net/models/transaction.py

```diff
@@ -1,28 +1,222 @@
 import warnings
-from typing import Sequence, Union
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional, Sequence, TypeVar, Union
 
-# noinspection PyPep8Naming
-from starkware.starknet.definitions.transaction_type import TransactionType as TT
-
-# noinspection PyPep8Naming
-from starkware.starknet.services.api.gateway.transaction import Declare as DCL
-from starkware.starknet.services.api.gateway.transaction import DeployAccount as DAC
-from starkware.starknet.services.api.gateway.transaction import InvokeFunction as IF
-from starkware.starknet.services.api.gateway.transaction import Transaction as T
+import marshmallow
+import marshmallow_dataclass
+from marshmallow import fields
+from starkware.starknet.services.api.contract_class import ContractClass
+from starkware.starknet.services.api.gateway.transaction_utils import (
+    compress_program_post_dump,
+    decompress_program,
+)
 
 from starknet_py.cairo.selector import get_selector_from_name
+from starknet_py.net.client_models import TransactionType
+from starknet_py.net.models import compute_address
 from starknet_py.net.models.chains import StarknetChainId
+from starknet_py.net.schemas.common import Felt, NoneFelt
 from starknet_py.utils.crypto.transaction_hash import (
     TransactionHashPrefix,
+    compute_declare_transaction_hash,
+    compute_deploy_account_transaction_hash,
     compute_transaction_hash,
 )
-from starknet_py.utils.docs import as_our_module
 
-Invoke = InvokeFunction = as_our_module(IF)
-Transaction = as_our_module(T)
-TransactionType = as_our_module(TT)
-Declare = as_our_module(DCL)
-DeployAccount = as_our_module(DAC)
+
+# pylint: disable=no-self-use, unused-argument
+@dataclass(frozen=True)
+class Transaction(ABC):
+    """
+    StarkNet transaction base class.
+    """
+
+    version: int = field(metadata={"marshmallow_field": Felt()})
+
+    @property
+    @abstractmethod
+    def type(self) -> TransactionType:
+        """
+        Returns the corresponding TransactionType enum.
+        """
+
+    @abstractmethod
+    def calculate_hash(self, chain_id: StarknetChainId) -> int:
+        """
+        Calculates the transaction hash in the StarkNet network - a unique identifier of the
+        transaction. See :ref:`compute_transaction_hash` docstring for more details.
+        """
+
+
+@dataclass(frozen=True)
+class AccountTransaction(Transaction):
+    """
+    Represents a transaction in the StarkNet network that is originated from an action of an
+    account.
+    """
+
+    max_fee: int = field(metadata={"marshmallow_field": Felt()})
+    signature: List[int] = field(
+        metadata={"marshmallow_field": fields.List(fields.String())}
+    )
+    nonce: int = field(metadata={"marshmallow_field": Felt()})
+
+
+# Used in the Account instead of Union[Invoke, Declare, DeployAccount]
+TypeAccountTransaction = TypeVar("TypeAccountTransaction", bound=AccountTransaction)
```

Comment:
Wouldn't AccountTransactionType be a better name?

Reply:
I think both are fine. Maybe `Type` and the beginning makes sense when the class is named TypeVar?

Reply:
I thought about `AccountTransactionType` too, but it's not a valid name for `TypeVar` 😞 
```
Pylint: Type variable name "AccountTransactionType" doesn't conform to predefined naming style
```